### ISSUE TITLE
Self-referential objects break decoders

### ISSUE THREAD

**Issue by sfarthin** [nvie/decoders#493]
If an object is passed in that refers to itself (like the express `req` object), an infinite loop occurs.

Smallest reproducible example:

```javascript
const { string } = require('decoders');

// Take any decoder and pass in some self-referential object
const req = {};
req.myself = req;

console.log(string(req)); // <--- RangeError: Maximum call stack size exceeded
```

**Comment by nvie** [comment]
Hey Steve, thanks for reporting this issue! This has been a known issue to me for quite some time. It's an implementation detail inside the underlying `debrief` library that I definitely want to fix. I'll see if I can take a look at this soon.

**Comment by sfarthin** [comment]
:+1: Thanks. I think even a quick-fix such as throwing a slightly more clear error message will go a long way.

For example, if the number of nested levels goes over 50, throw an error such as "Input is too large or self-referential" (maybe even with a link to this issue?). Default error messages are kinda broken when this large anyways.

**Comment by sfarthin** [comment]
Thanks for the good work on this library, its an essential tool when working with types in JS/TS.

**Comment by nvie** [comment]
> Default error messages are kinda broken when this large anyways.

If you want more succinct error messages, try using `guard(..., { style: 'simple' })` instead of just the default `guard(...)`. That will not echo back the input object, and just describe where the error is. That won't help you with this issue, but may help reduce your error messages in case of large objects.

**Comment by nvie** [comment]
This should now be fixed in [v1.20.0](https://github.com/nvie/decoders/releases/tag/v1.20.0). Please let me know if you're still seeing issues with this! Thanks for reporting. 🙏

**Comment by nvie** [comment]
This is the new behaviour under 1.20.0:

<img width="533" alt="Screen Shot 2020-05-25 at 13 55 40@2x" src="https://user-images.githubusercontent.com/83844/82810724-78a7db80-9e8f-11ea-8093-ad36628cd0bb.png">
